Updater vertical scaling

mikmaz opened the pull request.

zimnx commented on pkg/updater/updater.go:
```diff
@@ -24,71 +29,36 @@ func NewUpdater(c client.Client, logger log.Logger) Updater {
+		if equalChecksums, err := u.equalChecksums(ctx, sca); err != nil {
+			return err
+		} else if equalChecksums || u.recommendationExpired(ctx, sca) || !u.updateCooldownExceeded(ctx, sca) {
 			continue
```
Log on debug about why exactly recommendation wasn't applied

zimnx commented on pkg/updater/updater.go:
```diff
@@ -24,71 +29,36 @@ func NewUpdater(c client.Client, logger log.Logger) Updater {
 			return err
 		}
+		if !isScyllaClusterReady(cluster) {
+			u.logger.Error(ctx, "target readiness check", "sca", sca.Name, "namespace", sca.Namespace)
```
it's not error, info would be better

zimnx commented on pkg/updater/updater.go:
```diff
@@ -97,52 +67,161 @@ func (u *updater) RunOnce(ctx context.Context) error {
+func (u *updater) recommendationExpired(ctx context.Context, sca *v1alpha1.ScyllaClusterAutoscaler) bool {
+	if recExpTime := sca.Spec.UpdatePolicy.RecommendationExpirationTime; !sca.Status.LastUpdated.IsZero() &&
+		recExpTime != nil && time.Now().Sub(sca.Status.LastUpdated.Time) > recExpTime.Duration {
+		u.logger.Error(ctx, "sca's recommendation expired", "sca", sca.Name, "namespace", sca.Namespace)
```
error->info

zimnx commented on pkg/updater/updater.go:
```diff
@@ -97,52 +67,161 @@ func (u *updater) RunOnce(ctx context.Context) error {
+func (u *updater) updateCooldownExceeded(ctx context.Context, sca *v1alpha1.ScyllaClusterAutoscaler) bool {
+	if updateCooldown := sca.Spec.UpdatePolicy.UpdateCooldown; !sca.Status.LastApplied.IsZero() &&
+		updateCooldown != nil && time.Now().Sub(sca.Status.LastApplied.Time) < updateCooldown.Duration {
+		u.logger.Error(ctx, "update cooldown not exceeded", "sca", sca.Name, "namespace", sca.Namespace)
```
ditto

zimnx commented on pkg/updater/updater.go:
```diff
@@ -97,52 +67,161 @@ func (u *updater) RunOnce(ctx context.Context) error {
+}
+
+func (u *updater) updateSCAStatus(ctx context.Context, sca *v1alpha1.ScyllaClusterAutoscaler) error {
+	sca.Status.LastApplied = metav1.NewTime(time.Now())
```
always use UTC time, it will be stable during winter/summer time changes

zimnx commented on pkg/updater/updater.go:
```diff
@@ -97,52 +67,161 @@ func (u *updater) RunOnce(ctx context.Context) error {
+
+	if marshalledRecs, err := json.Marshal(sca.Status.Recommendations); err != nil {
+		return err
+	} else {
```
there's no need to have else block

zimnx commented on pkg/updater/updater.go:
```diff
@@ -97,52 +67,161 @@ func (u *updater) RunOnce(ctx context.Context) error {
+		return err
+	} else {
+		newChecksum := md5.Sum(marshalledRecs)
+		sca.Status.LastAppliedRecChecksum = &newChecksum
```
I thought about keeping checksum as a label in ScyllaCluster, not as a field in sca.Status. If ScyllaCluster update succeeds but this one fails, you will apply recommendation twice.

zimnx commented on pkg/updater/updater.go:
```diff
@@ -97,52 +67,161 @@ func (u *updater) RunOnce(ctx context.Context) error {
+	if marshalledRecs, err := json.Marshal(sca.Status.Recommendations); err != nil {
+		return err
+	} else {
+		newChecksum := md5.Sum(marshalledRecs)
```
use sha512 and extract function calculating hash for generic object to separate package

zimnx commented on pkg/updater/updater.go:
```diff
@@ -96,52 +71,172 @@ func (u *updater) RunOnce(ctx context.Context) error {
+func (u *updater) recommendationExpired(ctx context.Context, sca *v1alpha1.ScyllaClusterAutoscaler) bool {
+	if recExpTime := sca.Spec.UpdatePolicy.RecommendationExpirationTime; !sca.Status.LastUpdated.IsZero() &&
+		recExpTime != nil && time.Now().Sub(sca.Status.LastUpdated.Time) > recExpTime.Duration {
+		u.logger.Info(ctx, "sca's recommendation expired", "sca", sca.Name, "namespace", sca.Namespace)
```
I would still move logging outside of these predicates and instead writing `if (true) return true else return false` i would simply return if condition.

zimnx commented on pkg/updater/updater.go:
```diff
@@ -96,52 +71,172 @@ func (u *updater) RunOnce(ctx context.Context) error {
+	sca *v1alpha1.ScyllaClusterAutoscaler) (bool, error) {
+	if labels := cluster.ObjectMeta.Labels; labels != nil && sca.Status.Recommendations != nil {
+		if latestChecksum, ok := labels["sca-latest-checksum"]; ok {
+			newChecksum, err := util.NewChecksum(*sca.Status.Recommendations)
```
wouldn't it be easier if util.NewChecksum would return string?

zimnx commented on pkg/util/util.go:
```diff
@@ -57,3 +59,12 @@ func DurationPtr(value int) *metav1.Duration {
+func NewChecksum(obj interface{}) ([]byte, error) {
+	marshalledObj, err := json.Marshal(obj)
+	if err != nil {
+		return []byte{}, err
```
`return nil, err`

zimnx reviewed (commented):
minor comments

mikmaz committed:
Updater vertical scaling

- Updated filtering SCAs.
- Added readiness check of SC.
- Added vertical scaling support.
- Extended SCA with recommendation expiration check.
- Extended SCA with update cooldown.
- Added recommendations checksums as labels in SC.
- Added function calculating SHA 512 hash to utils.
- Extended unit tests.
- Refactored updater logic and tests.

zimnx reviewed (approved):
LGTM, thanks
